### ISSUE TITLE
UnnecessaryParentheses: false negative

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#11061]
https://checkstyle.org/config_coding.html#UnnecessaryParentheses

````
$ cat TestClass.java
public class TestClass {
     public boolean isUsed() {
        return used
                || (this.filePattern != null && this.filePattern.trim().length() == 0 // line 4
                || this.messagePattern != null && this.messagePattern.trim().length() == 0
                || this.checkPattern != null && this.checkPattern.trim().length() == 0);
     }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="UnnecessaryParentheses"/>
    </module>
</module>

$ java -jar checkstyle-9.2-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

I am expecting a violation on line 4 since the code equates to the same without the parenthesis as with.

Proof:
````
public class BooleanTest {

	public static void main(String[] args) {
		for (boolean a : new boolean[] { true, false }) {
			for (boolean b : new boolean[] { true, false }) {
				for (boolean c : new boolean[] { true, false }) {
					for (boolean d : new boolean[] { true, false }) {
						for (boolean e : new boolean[] { true, false }) {
							final boolean one = a || (b && c || d && e);
							final boolean two = a || b && c || d && e;

							if (one != two) {
								throw new IllegalStateException(a + ", " + b + ", " + c + ", " + d + ", " + e);
							}
						}
					}
				}
			}
		}

		System.out.println("Yes");
	}

}

Outputs:
"Yes"
````

**Comment by romani** [comment]
without parentheses it will be - yes, same:
```
return used 
         || this.filePattern != null && this.filePattern.trim().length() == 0
         || this.messagePattern != null && this.messagePattern.trim().length() == 0
         || this.checkPattern != null && this.checkPattern.trim().length() == 0)
```

**Comment by Vyom-Yadav** [comment]
on it.

**Comment by nmancus1** [comment]
We should update documentation to reflect that this check now supports some operator precedence: https://checkstyle.sourceforge.io/config_coding.html#UnnecessaryParentheses

**Comment by rnveach** [comment]
@nmancus1 Please add suggestion to PR since it is started.

**Comment by strkkk** [comment]
fix is merged

@Vyom-Yadav thanks :)
